### PR TITLE
Expose last_login_context on admin User

### PR DESCRIPTION
**Pull request by drernie**
## Summary
- Add `UserLastLoginContext` to GraphQL schema and regenerate client
- Expose `User.last_login_context` on `quilt3.admin` with SSO provider, id-token payload, matched mapping indices, assigned/active roles, admin flag, and login timestamp
- Populated when SSO config has `store_last_login_context: true`

## Test plan
- [x] `pytest tests/test_admin_api.py` (163 passed)
- [x] Smoke-test against a stack with `store_last_login_context: true`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<h3>Greptile Summary</h3>

This PR adds `UserLastLoginContext` to the GraphQL schema and Python admin API, exposing SSO login metadata (provider, id-token payload, matched mapping indices, assigned/active roles, admin flag, login timestamp) on `User.last_login_context`. The field is nullable and defaults to `None` so existing callers are unaffected; it is only populated when `store_last_login_context: true` is set in the SSO config.

<h3>Confidence Score: 5/5</h3>

Safe to merge — changes are additive, backward-compatible, and covered by 163 passing tests.

All findings are P2: a missing null-context test case and an Any typing artefact from code generation. Neither affects runtime correctness on the changed path. The new field has a default of None, existing callers are unaffected, and the full test suite passes.

No files require special attention; test_admin_api.py could benefit from a lastLoginContext: null test case.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| api/python/quilt3/admin/types.py | Adds `UserLastLoginContext` pydantic dataclass and optional `last_login_context` field to `User`; default `None` ensures backward compatibility with existing callers. |
| api/python/quilt3/_graphql_client/fragments.py | Adds `UserLastLoginContextSelection` BaseModel and `UserSelectionLastLoginContext` subclass; `id_token_payload` typed as `Any` (matching ariadne-codegen's scalar fallback) while the admin type uses `dict`, creating a minor type-safety gap. |
| shared/graphql/schema.graphql | Adds nullable `lastLoginContext: UserLastLoginContext` to `User` and the new `UserLastLoginContext` type; all fields are non-nullable within the type, which is correct since the whole object is only present when the context was captured. |
| api/python/tests/test_admin_api.py | Adds fixtures and assertion coverage for a non-null `lastLoginContext`; missing a test case where the user exists but `lastLoginContext` is `null` (SSO feature not yet enabled or user hasn't SSO-logged-in). |
| api/python/quilt3-graphql/queries.graphql | Adds `UserLastLoginContextSelection` fragment and wires it into `UserSelection`; consistent with other fragment patterns in the file. |
| api/python/quilt3/_graphql_client/client.py | Mechanically regenerated: `UserLastLoginContextSelection` fragment injected into every inline query that uses `UserSelection`; changes are repetitive but correct. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant User as End User (SSO)
    participant Backend as Quilt Backend
    participant GQL as GraphQL API
    participant Client as Python Client (quilt3)
    participant Admin as quilt3.admin

    User->>Backend: SSO login (id-token from provider)
    Backend->>Backend: Evaluate SSO mapping rules
    Backend->>Backend: Store last_login_context (if store_last_login_context: true)

    Admin->>Client: admin.users.get(username)
    Client->>GQL: UserSelection fragment (includes lastLoginContext)
    GQL->>Backend: Resolve User.lastLoginContext
    Backend-->>GQL: UserLastLoginContext or null
    GQL-->>Client: JSON response
    Client->>Client: UserSelection.model_validate(data)
    Client->>Client: User.from_gql(gql) -> cls(**gql.model_dump())
    Client-->>Admin: User(last_login_context=UserLastLoginContext(...) | None)
```

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `api/python/tests/test_admin_api.py`, line 812-821 ([link](https://github.com/quiltdata/quilt/blob/97802d6f18106155e77d1bf024349a06a8be9745/api/python/tests/test_admin_api.py#L812-L821)) 

   <a href="#">P2</a> **Missing test for `last_login_context=None`**

   The parameterized `test_get_user` test only covers (a) a user with a fully-populated `lastLoginContext` and (b) a missing user (`data=None`). There is no test for a user who exists but has never done an SSO login since `store_last_login_context` was enabled, which would return `lastLoginContext: null` from the server.

   Adding that case would ensure the optional path in `User.from_gql` (and pydantic's dict→`UserLastLoginContext` coercion) is exercised when the field is `null`:

   ```python
   USER_NO_LOGIN_CONTEXT = {**USER, "lastLoginContext": None}
   EXPECTED_USER_NO_LOGIN_CONTEXT = admin.User(
       **{**EXPECTED_USER.__dict__, "last_login_context": None}
   )

   @pytest.mark.parametrize(
       "data,result",
       [
           (USER, EXPECTED_USER),
           (USER_NO_LOGIN_CONTEXT, EXPECTED_USER_NO_LOGIN_CONTEXT),
           (None, None),
       ],
   )
   def test_get_user(data, result): ...
   ```

</details>

<sub>Reviews (1): Last reviewed commit: ["Update CHANGELOG for last\_login\_context"](https://github.com/quiltdata/quilt/commit/97802d6f18106155e77d1bf024349a06a8be9745) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=28794233)</sub>

> Greptile also left **1 inline comment** on this PR.

### CODE REVIEW

**Comment by drernie** [comment]
Live-tested against `tf-dev-unstable` (us-east-2):

```
Found 2 users
With last_login_context: 1
  ernest@quilt.bio: provider=okta active_role=public is_admin=True login_at=2026-04-17 16:18:14.418874+00:00
```

Schema, client codegen, and Python dataclass round-trip correctly end-to-end.

File: api/python/quilt3/_graphql_client/fragments.py [hunk 1]
```diff
@@ -126,6 +126,16 @@ class UnmanagedRoleSelection(BaseModel):
     arn: str
 
 
+class UserLastLoginContextSelection(BaseModel):
+    sso_provider: str = Field(alias="ssoProvider")
+    id_token_payload: Any = Field(alias="idTokenPayload")
```
[the remark this answers is not in the document]
Comment (drernie) [reply api/python/quilt3/_graphql_client/fragments.py]: Good catch — addressed in 0d0a659e by mapping the `JsonRecord` scalar to `dict` in the ariadne-codegen config. Generated code now types `id_token_payload` (and other `JsonRecord` selections) as `dict` rather than falling back to `Any`, so Pydantic rejects non-dict values at the client boundary instead of the admin-type boundary.

File: docs/CHANGELOG.md [hunk 2]
```diff
@@ -16,6 +16,12 @@ Entries inside each section should be ordered by type:
 
 # Changelog
 
+## unreleased
+
+### Python API
+
+* [Added] `quilt3.admin.User.last_login_context` exposes the most recent SSO login context (provider, id-token payload, matched mapping indices, assigned/active roles, admin flag, login timestamp); populated when SSO config has `store_last_login_context: true`
```
[the remark this answers is not in the document]
Comment (drernie) [reply docs/CHANGELOG.md]: Fixed in c1ea4380 — updated the header to `## unreleased - 2026-04-17`.

File: api/python/tests/test_admin_api.py [hunk 3]
```diff
@@ -97,6 +117,7 @@
     is_service=False,
     role=EXPECTED_UNMANAGED_ROLE,
     extra_roles=[EXPECTED_MANAGED_ROLE],
+    last_login_context=EXPECTED_LAST_LOGIN_CONTEXT,
 )
```
[the remark this answers is not in the document]
Comment (drernie) [reply api/python/tests/test_admin_api.py:122]: Fixed in c1ea4380 — `test_get_user` now parametrizes over both the populated and `lastLoginContext: null` cases (plus the existing `None` user case).

File: api/python/quilt3/_graphql_client/fragments.py [hunk 4]
```diff
@@ -126,6 +126,16 @@ class UnmanagedRoleSelection(BaseModel):
...
 
 
+class UserLastLoginContextSelection(BaseModel):
+    sso_provider: str = Field(alias="ssoProvider")
+    id_token_payload: dict = Field(alias="idTokenPayload")
+    matched_mapping_indices: List[int] = Field(alias="matchedMappingIndices")
+    assigned_roles: List[str] = Field(alias="assignedRoles")
+    active_role: str = Field(alias="activeRole")
+    is_admin: bool = Field(alias="isAdmin")
+    login_at: datetime = Field(alias="loginAt")
```
[the remark this answers is not in the document]
Comment (drernie) [reply api/python/quilt3/_graphql_client/fragments.py:136]: Addressed in 0d0a659e by adding a `JsonRecord` scalar mapping in the ariadne-codegen config; generated `id_token_payload` is now typed as `dict` instead of falling back to `Any`.

File: api/python/quilt3/admin/types.py [hunk 5]
```diff
@@ -79,6 +79,23 @@ def from_gql(cls, gql: _graphql_client.BaseModel) -> "Policy":
...
+@pydantic.dataclasses.dataclass
+class UserLastLoginContext:
+    """Details about a user's most recent SSO login.
+
+    Populated only when the SSO config has ``store_last_login_context: true``
+    and the user has logged in via SSO at least once since it was enabled.
+    """
+
+    sso_provider: str
+    id_token_payload: dict
```
[the remark this answers is not in the document]
Comment (drernie) [reply api/python/quilt3/admin/types.py]: Fixed in c1ea4380 — `id_token_payload` is now `dict[str, typing.Any]`.

**Comment by sir-sigurd** [comment]
do we still need this?

**Comment by drernie** [comment]
No, sorry, it is handled elsewhere.
